engine: resources: Add a WatchFiles func to exec

lourencovales opened the pull request:
This adds a function that keeps track of changes to a given file inside the exec resource.

Keep this as a draft PR for now because it seems like we're generating more events than we should, looking for some feedback into this.

purpleidea commented on engine/resources/exec.go:
```diff
@@ -119,6 +119,9 @@ type ExecRes struct {
 	WatchCwd string `lang:"watchcwd" yaml:"watchcwd"`
 
+	// WatchFile is the file that the WatchFile routine will track.
+	WatchFile string `lang:"watchfile" yaml:"watchfile"`
```
Don't we want this to be a list of files?

I have a hint to implementing this if you feel stuck.

purpleidea commented on engine/resources/exec.go:
```diff
@@ -281,6 +285,45 @@ func (obj *ExecRes) Watch(ctx context.Context) error {
+				}
+
+				select {
+				case rwChan <- event:
```
You may not want to piggyback this onto the existing rwChan...

purpleidea reviewed (commented):
Looks great, let's change it up a bit though =D

lourencovales commented:
@purpleidea this should in better shape now

purpleidea commented on engine/resources/exec.go:
```diff
@@ -281,6 +286,46 @@ func (obj *ExecRes) Watch(ctx context.Context) error {
 	}
 
+	for _, file := range obj.WatchFiles {
+		// since we're only dealing with files, recurse can be set to false
```
Why can't we be dealing with directories? From a API/UI point of view, how would we let the user differentiate between the two?

lourencovales replied:
I thought this was the point of the function - to only deal with files. I can implement the ability for taking a directory as argument if that's a desired feature.
As for the other question, a directory would have to end with a forward slash (/), while a file wouldn't.

purpleidea replied:
Correct =D

purpleidea commented on engine/resources/exec.go:
```diff
@@ -281,6 +286,46 @@ func (obj *ExecRes) Watch(ctx context.Context) error {
+		}
+		defer recWatcher.Close()
+
+		obj.wg.Add(1)
```
There's a bit of an issue that needs investigating...

_We_ definitely want to use a local wg. Not an obj.wg and it should have the same lifetime as the Watch function.

But there's some legacy code in obj.wg which we should see if we (probably I) ported correctly or is a bug/unoptimal scenario from years ago.

At least for us, we don't want obj.wg 

Do you understand why?

lourencovales replied:
If we're using obj.wg we're using a shared the same wait group instance across goroutines. Would that be a source of funkyness? If so, we definitely want a local wg.

purpleidea replied:
We want the local wg, since it's lifetime is only limited to the Watch function. If we had a helper function watch called, then we'd want something shared as well, but we don't here.

purpleidea reviewed (commented):
As you may have already figured, we probably want this to look similar to the: https://github.com/purpleidea/mgmt/blob/master/engine/resources/file.go#L477 code

lourencovales commented:
> Looks really nice, some small nits to look into. The most important question, does this feel natural to you? Does it make sense?

The whole "dance" around go routines became much more understandable after working on this issue, that's for sure. That's my biggest takeaway from this PR.

lourencovales commented:
@purpleidea we should be good to go now!

lourencovales commented:
@purpleidea this should in better shape now, let me know if there's anything missing

purpleidea commented:
I see a bug. What if Watchfiles of one res is len 4 and the other is 3?

lourencovales commented:
> I see a bug. What if Watchfiles of one res is len 4 and the other is 3?

@purpleidea I fixed this, and I also added another check in the Validate function that I noticed was also missing. Let me know if you spot anything else

purpleidea commented:
Thank you!

I merged a slightly retouched version. Have a look at the diff =D

lourencovales commented:
Thank you for merging the patch. I'll keep in mind the engineUtil.StrListCmp util :-)

purpleidea commented:
> Thank you for merging the patch. I'll keep in mind the engineUtil.StrListCmp util :-)

Thank _you_ for the patch! (I forget about that util too.) But note the change I made in Validate.
